.github/workflows/build-and-publish.yml

```diff
@@ -0,0 +1,508 @@
+          Target PEP440-compliant version to release.
+          Please, don't prepend `v`.
+        required: true
+        type: string
+      release-commitish:
+        # github.event_name == 'workflow_dispatch'
+        # && github.event.inputs.release-commitish
+        default: ''
+        description: >-
+          The commit to be released to PyPI and tagged
+          in Git as `release-version`. Normally, you
+          should keep this empty.
+        type: string
+
+concurrency:
+  group: >-
+    ${{
+        github.workflow
+    }}-${{
+        github.ref_type
+    }}-${{
+        github.event.pull_request.number || github.sha
+    }}
+  cancel-in-progress: true
+
+env:
+  DEFAULT_PYTHON_VERSION: 3.11
+  FORCE_COLOR: 1  # Request colored output from CLI tools supporting it
+  PIP_DISABLE_PIP_VERSION_CHECK: 1
+  PIP_NO_PYTHON_VERSION_WARNING: 1
+  PIP_NO_WARN_SCRIPT_LOCATION: 1
+
+permissions: {}  # Jobs hardened by default
+
+jobs:
+  pre-setup:
+    name: ⚙️  Pre-set global build settings
+    runs-on: ubuntu-latest
+
+    outputs:
+      dist-version: ${{ github.event.inputs.release-version }}
+      git-tag: ${{ steps.git.outputs.tag }}
+      stable-branch: stable-${{ steps.version.outputs.major }}
+      release-branch: ${{ steps.git.outputs.release-branch }}
+      sdist-artifact-name: ${{ steps.artifact-name.outputs.sdist }}
+      changelog-patch-name: ${{ steps.changelog-patch-name.outputs.filename }}
+      changelog-draft-name-md: >-
+        ${{ steps.changelog-draft-name.outputs.filename-base }}.md
+      changelog-draft-name-rst: >-
+        ${{ steps.changelog-draft-name.outputs.filename-base }}.rst
+
+    permissions:
+      contents: read  # For accessing the repository
+
+    steps:
+    - name: >-
+        ✅ Fail if the requested release version
+        '${{ github.event.inputs.release-version }}' is not PEP440-compliant
+      run: |
+        import os
+        import pathlib
+        import sys
+
+        from packaging.version import InvalidVersion, Version
+
+        try:
+            Version('${{ github.event.inputs.release-version }}')
+        except InvalidVersion as version_error:
+            with pathlib.Path(
+                os.environ['GITHUB_STEP_SUMMARY'],
+            ).open('a') as summary_fd:
+                print('# ❌ Requested version check failed', file=summary_fd)
+                print('', file=summary_fd)
+                print(
+                    'The requested version `${{
+                        github.event.inputs.release-version
+                    }}` must be compliant with PEP440 but it is not.',
+                    file=summary_fd,
+                )
+                print('', file=summary_fd)
+                print(str(version_error), file=summary_fd)
+                print('', file=summary_fd)
+
+            sys.exit(str(version_error))
+      shell: python
+
+    - name: ⚙️  Set the target Git 🏷️ tag
+      id: git
+      run: |
+        echo 'release-branch=release/${{
```

Comment:
@mkrizek if we can make ansibot allow PRs from in-upstream release branches, I would also make this workflow create PRs, not just a branch. WDYT?

Reply:
https://github.com/ansible/ansibullbot/issues/971 from @mattclay is the original issue for the bot requesting to close PRs from an upstream branch and cancel the CI run for it. I am not sure we want to re-allow such PRs though I don't remember why we disallowed them in the first place :-)

I suppose at the very least if we wanted to we could make an exception for PRs from this workflow (based on criteria like PR submitter and/or branch name, $other) 🤷🏻‍♂️

Reply:
> [ansible/ansibullbot#971](https://github.com/ansible/ansibullbot/issues/971) from @mattclay is the original issue for the bot requesting to close PRs from an upstream branch and cancel the CI run for it. I am not sure we want to re-allow such PRs though I don't remember why we disallowed them in the first place :-)

This is because when the upstream is populated with random branches, the end-users will get them copied in forks. Depending on the time, different users would have a copy of such branches in different states, and they won't be cleaned up from the forks once unnecessary. Plus, many branches in forks would create confusion too. So in general, it's a good idea to keep the upstream clean, except for special cases such as this one.

> I suppose at the very least if we wanted to we could make an exception for PRs from this workflow (based on criteria like PR submitter and/or branch name, $other) 🤷🏻‍♂️

Yes, that's exactly what I'm asking — making an exception for certain PR. Either by `user.type == 'Bot'` or by some label. The bot check would work for any GitHub App (if we even have one creating PRs)  and the label could take care of the human-like bots. The branch pattern could work but would require re-hardcoding every time with add some other automation needing this.

Reply:
There were two reasons for not wanting upstream branches:

1) Branch pollution, as @webknjaz mentioned.
2) Duplicate CI runs for PRs (one from the branch, another from the PR). This is no longer an issue, now that we're using AZP and have configured it to only test specific branches.

@webknjaz Is it not possible for the workflow to create branches in a user's fork instead?

Reply:
@mattclay the auto-provisioned `GITHUB_TOKEN` secret is very limited in what it can do. For example, when it mutates anything on the platform, the corresponding events don't propagate to any listeners (neither other apps/integrations, not GHA itself, including other job definitions).

But yes, It is possible with extra maintenance burden being the following:
0) a (dedicated?) user/org with a fork
1) somebody with full repo access will have to maintain a secret with an API key for accessing another user resources — IIRC they now have a mandatory expiration date
2) there needs to be a guarantee that the fork's branches aren't going to be overridden

I suppose, it could be easier to have GitHub App credentials because of (1). We'd still need to have a dedicated forked repository, maybe under one of our orgs like community it'd work. That App would need to be installed here and in the fork (this bit is easy). There's no mandatory need to actually run a web server for the app listening to the events — the secrets could be just used in GHA directly for getting a qualifying token.

This is one of the [Patchback](https://github.com/apps/patchback)'s annoyances — many people do want backport PRs from forks. But then, it's harder to maintain such a system. This is mostly solved by enabling the `Automatically delete head branches` checkbox in the repository settings, which lets GitHub auto-clean the corresponding branches post-merge.

Now that I think about it more, it would definitely be useful to have a GHA regardless of whether we'll accept the burden of mangling with forks.

P.S. If we were to have a more canonical type of GitHub App with a web-app interface, we could authenticate users and just use their forks w/o having a dedicated one. But that would be substantially more work, and it's probably not worth it.

Reply:
> when it mutates anything on the platform, the corresponding events don't propagate to any listeners

This is why the current implementation doesn't attempt to auto-create a PR.